File: apps/ios/BrickBack/Features/Catalog/SetMinifigsScreen.swift
import SwiftUI
import BrickBackKit

/// `.setMinifigs` — the minifigs that belong to a set, read from the catalog. Preview only;
/// minifig verification happens in the review flow (S4). Port of `set_minifigs_screen.dart`.
struct SetMinifigsScreen: View {
    @Environment(AppEnvironment.self) private var env
    @Environment(\.activeRouter) private var activeRouter
    let itemId: Int

    @State private var state: LoadState<[CatalogMinifig]> = .loading

    private var router: Router { activeRouter ?? env.homeRouter }

    /// On iOS 18+ this screen rides the native nav bar (back chevron + inline title), matching the
    /// set-detail screen it's pushed from and the rest of the catalog family. The iOS 17 legacy
    /// pushed flow has no reliable native bar, so it keeps its `ScreenHeader`.
    private var systemProvidesBack: Bool {
        if #available(iOS 18.0, *) { return true }
        return activeRouter === env.searchRouter
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            if !systemProvidesBack {
                ScreenHeader(L.minifigs, onBack: { router.pop() })
            }
            switch state {
            case .idle, .loading:
                ProgressView().tint(AppColors.primary).frame(maxWidth: .infinity, maxHeight: .infinity)
            case .failed(let message):
                EmptyState(title: L.minifigsCouldntLoad, message: message, icon: "exclamationmark.triangle")
            case .loaded(let minifigs):
                if minifigs.isEmpty {
                    EmptyState(
                        title: L.minifigsEmptyTitle,
                        message: L.minifigsEmptyMessage,
                        icon: "person"
                    )
                } else {
                    ScrollView {
                        LazyVStack(spacing: AppSpacing.s8) {
                            ForEach(minifigs, id: \.minifigItemId) { fig in MinifigRow(minifig: fig) }
                        }
                        .padding(.horizontal, AppSpacing.screen)
                        .padding(.top, AppSpacing.s8)
                        .padding(.bottom, AppSpacing.s24)
                    }
                }
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(AppColors.canvas)
        // Only shown when the native bar is present (iOS 18+); harmless where it's hidden.
        .navigationTitle(L.minifigs)
        .navigationBarTitleDisplayMode(.inline)
        .task(id: itemId) {
            state = .loading
            do {
                let figs = try await env.services.catalog.setMinifigs(itemId)
                if !Task.isCancelled { state = .loaded(figs) }
            } catch {
                if !Task.isCancelled { state = .failed("\(error)") }
            }
        }
    }
}

private struct MinifigRow: View {
    let minifig: CatalogMinifig

    var body: some View {
        AppCard(padding: AppSpacing.s12) {
            HStack(spacing: AppSpacing.s12) {
                SetThumb(imageUrl: minifig.imageUrl, size: 44, label: "🧍")
                VStack(alignment: .leading, spacing: 2) {
                    Text(minifig.name).font(AppText.body).foregroundStyle(AppColors.ink).lineLimit(1)
                    if !minifig.figNum.isEmpty {
                        Text(minifig.figNum).font(AppText.caption).foregroundStyle(AppColors.inkSoft)
                    }
                }
                Spacer(minLength: AppSpacing.s8)
                Text("×\(minifig.quantity)").font(AppText.label).foregroundStyle(AppColors.ink)
            }
        }
    }
}

File: apps/mobile/lib/features/catalog/set_minifigs_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../../l10n/l10n.dart';
import '../../theme/app_theme.dart';
import '../../widgets/primitives.dart';
import 'catalog_models.dart';
import 'catalog_repository.dart';

/// `/set/:id/minifigs` — the minifigs that belong to a set, read from the
/// catalog. Preview only; minifig verification happens in the review flow.
class SetMinifigsScreen extends ConsumerWidget {
  const SetMinifigsScreen({super.key, required this.itemId});

  final int itemId;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final c = BrickColors.of(context);
    final minifigs = ref.watch(setMinifigsProvider(itemId));
    return Scaffold(
      body: SafeArea(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            ScreenHeader(context.l10n.minifigs, onBack: () => Navigator.of(context).maybePop()),
            Expanded(
              child: minifigs.when(
                loading: () =>
                    Center(child: CircularProgressIndicator(color: c.primary)),
                error: (e, _) => EmptyState(
                  icon: Icons.error_outline,
                  title: context.l10n.minifigsCouldntLoad,
                  message: '$e',
                ),
                data: (list) {
                  if (list.isEmpty) {
                    return EmptyState(
                      icon: Icons.person_outline,
                      title: context.l10n.minifigsEmptyTitle,
                      message: context.l10n.minifigsEmptyMessage,
                    );
                  }
                  return ListView.separated(
                    padding: const EdgeInsets.fromLTRB(
                        AppSpacing.screen, AppSpacing.s8, AppSpacing.screen, AppSpacing.s24),
                    itemCount: list.length,
                    separatorBuilder: (_, _) => const SizedBox(height: AppSpacing.s8),
                    itemBuilder: (context, i) => _MinifigRow(minifig: list[i]),
                  );
                },
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class _MinifigRow extends StatelessWidget {
  const _MinifigRow({required this.minifig});
  final CatalogMinifig minifig;

  @override
  Widget build(BuildContext context) {
    return AppCard(
      padding: const EdgeInsets.all(AppSpacing.s12),
      child: Row(
        children: [
          SetThumb(imageUrl: minifig.imageUrl, size: 44, label: '🧍'),
          const SizedBox(width: AppSpacing.s12),
          Expanded(
            child: Column(
              crossAxisAlignment: CrossAxisAlignment.start,
              children: [
                Text(minifig.name,
                    maxLines: 1, overflow: TextOverflow.ellipsis, style: AppText.body),
                if (minifig.figNum.isNotEmpty) ...[
                  const SizedBox(height: 2),
                  Text(minifig.figNum, style: AppText.caption),
                ],
              ],
            ),
          ),
          const SizedBox(width: AppSpacing.s8),
          Text(context.l10n.quantityTimes(minifig.quantity), style: AppText.label),
        ],
      ),
    );
  }
}
